observability/opentelemetry/v2/README.md

```diff
@@ -0,0 +1,186 @@
+```
+
+With the above configuration, the spans generated by the `OTelObservabilityService` will have:
+
+<table>
+  <tbody>
+    <tr>
+      <td style="font-weight:bold">Span name</td>
+      <td><code>cloudevents.client.[event type] [operation name]</code> where <code>[operation name]</code> follows the OpenTelemetry <a href="https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/trace/semantic_conventions/messaging.md#operation-names">semantic conventions for messaging systems</a></td>
+    </tr>
+    <tr>
+      <td style="font-weight:bold">Span attributes</td>
+      <td>The attributes as defined in <a href="https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/observability/keys.go">keys.go</a></td>
+    </tr>
+    </tbody>
+</table>
+
+If you require different span names or extra attributes, you can pass multiple `OTelObservabilityServiceOption` options when creating the observability service:
+
+```go
+nameFormatter := func(e *cloudevents.Event) string {
+	return "my.custom.name." + e.Context.GetType()
+}
+
+attributesGetter := func(*cloudevents.Event) []attribute.KeyValue {
+	return []attribute.KeyValue{
+		attribute.String("my-attr", "some-value"),
+	}
+}
+
+// create the obs service with custom span names and attributes
+os := otelObs.NewOTelObservabilityService(
+	otelObs.WithSpanNameFormatter(nameFormatter),
+	otelObs.WithSpanAttributesGetter(attributesGetter),
+)
+
+c, err := cloudevents.NewClient(p, client.WithObservabilityService(os))
+```
+
+>NOTE: The `nameFormatter` and `attributesGetter` functions will be called on each span creation. **Avoid** doing any heavy processing in them.
+
+
+## Extra types
+
+This package also contains extra types and helper functions that are useful in case you need to access/set the `tracecontext` in a more "low-level" way.
+
+They allow to inject(write) and extract(read) `tracecontext` from the event. This is particularly useful when dealing with code that has no notion of a "request" nor a context. For example, long-running background processes pooling from a queue. 
+
+>Note: To learn more about the propagation, take a look at the [Propagators API SPEC](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/context/api-propagators.md).
+
+
+## Manually extracting/injecting tracecontext from the event
+
+When working with distributed systems, it can be difficult to achieve proper context propagation. For example, a long-running process listening to a topic does not have a "context" concept like a HTTP server receiving requests does.
+
+>Note: The OpenTelemetry community is always creating auto-instrumentation libraries for popular SDKs and frameworks, so there could be one for the queue client or other framework you are using. It's worth checking out what is available in the [go-contrib GitHub repository](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation)
```

Comment:
AFAIK, the way cloudevents works (at least this one for go) is the `client.go` they offer has a transport that can be many things. HTTP, amqp, nats, kafka, etc. Since there's many protocols, I believe the idea with this Observability service is to always have spans being created when using the client.

I guess listing the Registry also could work. Since this is for Go, I thought of just putting the contrib link but maybe something else goes into the "core" one so it gets missed.

I added auto-instrumentation for HTTP, so incoming/outgoing requests have propagation working. If there are libraries for the other protocols, those could also be added later on for sure.

Reply:
I rephrased it a bit stating the registry and as support increases, new auto-instrumentation libraries can be added.